HashRegistrar.sol

```diff
@@ -0,0 +1,320 @@
+    calculated as ```averagePrice = averagePrice * 0.999 + newPrice * 0.001```**. The 
+    averagePrice at the moment of purchase is also registered on the contract.
+    */ 
+    function revealBid(bytes32 _hash, address _owner, uint _value, bytes32 _salt) noEther  {
+        bytes32 seal = shaBid(_hash, _owner, _value, _salt);
+        Deed bid = sealedBids[seal];
+        if (address(bid) == 0 ) throw;
+        bid.unsealBid(_hash, _owner, _value, _salt);
+        sealedBids[seal] = Deed(0);
+        
+        entry hash = hashes[bid.hash()];
+        
+        if (bid.creationDate() > hash.auctionExpiration - revealPeriod
+            || now > hash.auctionExpiration ) {
+            // bid is invalid, refund 99.9%
+            bid.closeDeed(1);
+            
+        } else if ( _value < averagePrice / 100 ) {
+            // bid is invalid but not punishable
+            bid.closeDeed(1000);
+            
+        } else if (_value > hash.highestBid) {
+            // new winner
+            // cancel the other bid, burn 0.1%
+            Deed previousWinner = Deed(hash.deed);
+            previousWinner.closeDeed(999);
+            
+            // set new winner
+            hash.value = hash.highestBid;
+            hash.highestBid = _value;
+            hash.deed = sealedBids[seal];
+        
+        } else if (_value > hash.value) {
+            // not winner, but affects second place
+            hash.value = _value;
+            bid.closeDeed(999);
+            
+        } else {
+            // bid doesn't affect auction
+            bid.closeDeed(999);
+        }
+    }
+    
+    function cancelBid(bytes32 seal) {
+        Deed bid = sealedBids[seal];
+        if (address(bid) == 0 || now < bid.creationDate() + auctionLength * 2 || bid.hash() > 0) throw; 
+        bid.closeDeed(0);
+        sealedBids[seal] = Deed(0);
+    }
+    
+    function finalizeAuction(bytes32 _hash) noEther {
+        entry hash = hashes[_hash];
+        if (now < hash.auctionExpiration 
+            || hash.value == 0
+            || hash.status != Mode.Auction) throw;
+        
+        // set the hash
+        hash.status = Mode.Owned;
+        hash.firstRegistered = now;
+        hash.lastRenewed = now;
+
+        //Calculate the moving average period as a way to measure frequency
+        uint period = (now - lastSinceNewRegistry) * M;
+        averagePeriod = (999 * averagePeriod + period) / 1000;
+        // 60 seconds is the average of 10,000 registrations per week. Use whatever's higher.
+        uint n = averagePeriod < (60 seconds * M) ? 7 days * M / averagePeriod : 9999;
+        averagePrice = (n*averagePrice + hash.value ) / (n+1);
+        hash.averagePrice = averagePrice;
+        lastSinceNewRegistry = now;
+        
+        Deed deedContract = Deed(hash.deed);
+        deedContract.setBalance(hash.value);
+    }
+    
+    /*
+    ## Renewals can be done at any moment by renewing the deposit
```

Comment:
I actually agreed with your reasoning that allowing people to unregister almost immediately made it too easy to submit bids to escalate prices.